blockchain/blockchain.go

```diff
@@ -1099,96 +1100,97 @@ type BlockSignFunc func(blockHash, stateDiffCommitment *felt.Felt) ([]*felt.Felt
 
 // Finalise will calculate the state commitment and block hash for the given pending block and append it to the
 // blockchain
-func (b *Blockchain) Finalise(pending *Pending, sign BlockSignFunc) error {
+func (b *Blockchain) Finalise(txs []core.Transaction, receipts []*core.TransactionReceipt, diff *core.StateDiff,
+	classes map[felt.Felt]core.Class, seqAddr *felt.Felt, sign BlockSignFunc,
+) error {
 	return b.database.Update(func(txn db.Transaction) error {
-		var err error
-
-		state := core.NewState(txn)
-		pending.StateUpdate.OldRoot, err = state.Root()
-		if err != nil {
-			return err
-		}
-
-		if h, hErr := chainHeight(txn); hErr == nil {
-			pending.Block.Number = h + 1
-		}
-
-		if err = state.Update(pending.Block.Number, pending.StateUpdate.StateDiff, pending.NewClasses); err != nil {
-			return err
-		}
-
-		pending.Block.GlobalStateRoot, err = state.Root()
-		if err != nil {
-			return err
-		}
-		pending.StateUpdate.NewRoot = pending.Block.GlobalStateRoot
-
-		var commitments *core.BlockCommitments
-		pending.Block.Hash, commitments, err = core.BlockHash(pending.Block, b.network)
+		pending, err := pendingBlock(txn)
 		if err != nil {
 			return err
 		}
-		pending.StateUpdate.BlockHash = pending.Block.Hash
 
-		var sig []*felt.Felt
-		sig, err = sign(pending.Block.Hash, pending.StateUpdate.StateDiff.Commitment())
+		// TODO we can avoid this disk read by storing the pending block's number
+		// in storeEmptyPending. That will require some refactoring in the rpc
+		// package to do correctly.
+		height, err := chainHeight(txn)
 		if err != nil {
 			return err
 		}
-		pending.Block.Signatures = [][]*felt.Felt{sig}
-
-		if err = b.storeBlock(txn, pending.Block, commitments); err != nil {
-			return err
+		pending.Block.Number = height + 1
+		pending.Block.SequencerAddress = seqAddr
+		pending.Block.Transactions = txs
+		pending.Block.TransactionCount = uint64(len(txs))
+		pending.Block.Receipts = receipts
+		for _, receipt := range receipts {
+			pending.Block.EventCount += uint64(len(receipt.Events))
 		}
+		pending.Block.EventsBloom = core.EventsBloom(receipts)
+		pending.Block.Timestamp = uint64(time.Now().Unix())
+		pending.Block.ProtocolVersion = supportedStarknetVersion.String()
+		// TODO gas prices
+		pending.StateUpdate.StateDiff = diff
```

Comment:
Do we not also have to set the NewRoot, OldRoot and BlockHash?

Reply:
Ah, nvm, just seen that it's set in `storeEmptyPending()` whenever we store a pending block.